src/json-observe.js

```diff
@@ -0,0 +1,189 @@
+/** Class representing a JS Object observer  */
+var JsonObserver = (function () {
+    function JsonObserver(root) {
+        this.originalObject = root;
+        this.cachedProxy = null;
+        this.isRecording = false;
+        this.userCallback;
+        var sender = this;
+        /*
+        instead of setting a boolean to
+        be checked every time an operation
+        is being recorded, I'm removing the
+        whole callback when it's disabled.
+        Because we are only disabling the callback
+        once in the beginning, no need to `if` every time
+        */
+        this.switchObserverOn = function () {
+            sender.defaultCallback = function (event) {
+                if (sender.isRecording) {
+                    sender.patches.push(event);
+                }
+                if (sender.userCallback) {
+                    sender.userCallback(event);
+                }
+            };
+        };
+        this.switchObserverOff = function () {
+            sender.defaultCallback = function () { };
+        };
+    }
+    JsonObserver.deepClone = function (obj) {
+        switch (typeof obj) {
+            case "object":
+                return JSON.parse(JSON.stringify(obj)); //Faster than ES5 clone - http://jsperf.com/deep-cloning-of-objects/5
+            case "undefined":
+                return null; //this is how JSON.stringify behaves for array items
+            default:
+                return obj; //no need to clone primitives
+        }
+    };
+    JsonObserver.escapePathComponent = function (str) {
+        if (str.indexOf('/') === -1 && str.indexOf('~') === -1)
+            return str;
+        return str.replace(/~/g, '~0').replace(/\//g, '~1');
+    };
+    JsonObserver.prototype.generateProxyAtPath = function (obj, path) {
+        if (!obj) {
+            return obj;
+        }
+        var instance = this;
+        var proxy = new Proxy(obj, {
+            get: function (target, propKey, receiver) {
+                if (propKey.toString() === '_proxy') {
```

Comment:
I think it's worth to describe side-effect of `_proxy` hack in readme. If user object contains `_proxy` property it will not get recursively proxified.

Maybe we could advertise it as a feature?
Rename it to `_dontProxify` so we will ~"set" it in a getter during recursive proxification, but users may set a real property to block proxification/observing on some level?